src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -504,6 +551,149 @@ def time_remaining(self) -> Optional[float]:
+
+    For the `__await__` method is it is proxied to the intercepted call only when
+    the interceptor task is finished.
+    """
+
+    _loop: asyncio.AbstractEventLoop
+    _channel: cygrpc.AioChannel
+    _write_to_iterator_async_gen: Optional[AsyncIterable[RequestType]]
+    _write_to_iterator_queue: Optional[asyncio.Queue]
+
+    _FINISH_ITERATOR_SENTINEL = tuple()
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, interceptors: Sequence[StreamUnaryClientInterceptor],
+                 request_iterator: Optional[RequestIterableType],
+                 timeout: Optional[float], metadata: MetadataType,
+                 credentials: Optional[grpc.CallCredentials],
+                 wait_for_ready: Optional[bool], channel: cygrpc.AioChannel,
+                 method: bytes, request_serializer: SerializingFunction,
+                 response_deserializer: DeserializingFunction,
+                 loop: asyncio.AbstractEventLoop) -> None:
+        self._loop = loop
+        self._channel = channel
+        if not request_iterator:
+            # We provide our own request iterator which is a proxy
+            # of the futures writes that will be done by the caller.
+            self._write_to_iterator_queue = asyncio.Queue(maxsize=1)
+            self._write_to_iterator_async_gen = self._proxies_writes_as_a_request_iteerator(
+            )
+            request_iterator = self._write_to_iterator_async_gen
+        else:
+            self._write_to_iterator_queue = None
+
+        interceptors_task = loop.create_task(
+            self._invoke(interceptors, method, timeout, metadata, credentials,
+                         wait_for_ready, request_iterator, request_serializer,
+                         response_deserializer))
+        super().__init__(interceptors_task)
+
+    # pylint: disable=too-many-arguments
+    async def _invoke(
+            self, interceptors: Sequence[StreamUnaryClientInterceptor],
+            method: bytes, timeout: Optional[float],
+            metadata: Optional[MetadataType],
+            credentials: Optional[grpc.CallCredentials],
+            wait_for_ready: Optional[bool],
+            request_iterator: RequestIterableType,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> StreamUnaryCall:
+        """Run the RPC call wrapped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryUnaryClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request_iterator: RequestIterableType
+        ) -> _base_call.StreamUnaryCall:
+
+            interceptor = next(interceptors, None)
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+
+                return await interceptor.intercept_stream_unary(
+                    continuation, client_call_details, request_iterator)
+            else:
```

Comment:
nit: Should we support interceptors to provide response object directly like unary call interceptor? If not, we might want to document it.

Reply:
I thought about that, but I did not provide support for it due to the nature of the RPC which expects a stream of objects during the request stage.

IMO the interceptor would need to decide what object is being returned considering all of the requests that re being sent, and unfortunately the interceptor - as it is being designed right now- would not have the way of returning an alternative response in that specific context. Why?  Despite the interceptors are able to observe all of the requests that are sent by the caller, at that specific moment the interceptor already returned the call.

Reply:
If we allow the interceptor to return a finished call, the user would need to consider to implement a defensive pattern every time that the `write` is used, for example:

```python
cal = StreamingUnaryCall()

try:
    for message in messages:
        await call.write(message)
except InvalidStateError:
    # Is this because an error or because there is a response?
    pass

try:
    response = await call
except (AioRpcError, asyncio.CancelledError):
    # there was an error
```

Also, Im wondering how the interceptor would now need to be returned if the interceptor does not have any chance of consuming the messages? should be the response to something that is built based on what has been observed in the message requests?

Having the feeling that we would need to have a very good example of implementing what you are saying, should we wait till the community askes for this?

Reply:
I think you are right, we can wait till the community ask for this.

Sending messages or reading messages from wire is not guaranteed to success. Applications need to handle error cases inevitably.

In the snippet, the interceptor pre-maturely terminated the call and cause following message sending to be failed. This can be valid case, for example, what if the interceptor detects some error among the request messages like checksum mismatch, then the interceptor abort the call.